### PR TITLE
Enforce cognitive-complexity limit everywhere; narrow scripts biome override

### PR DESCRIPTION
**Pull request by stefan-burke**
## Summary

Removes biome overrides that let parts of the codebase off the hook, and refactors the code that then failed to lint.

## Changes

**`biome.json`**
- Replaced the blanket `scripts/**/*.ts` override (which disabled `noExcessiveCognitiveComplexity`, `noAssignInExpressions`, `noConsole`, and `noImplicitAnyLet`) with one that only keeps `noConsole` off — console output is intentional in CLI/build/test scripts. (`noAssignInExpressions` and `noImplicitAnyLet` are `off` at the top level, so they remain off everywhere; this override no longer redundantly re-disables them.)
- Removed the `test/**` override that raised `noExcessiveCognitiveComplexity` to max 30. Tests now use the base max of 15.

Net effect: `noExcessiveCognitiveComplexity` (max 15) is now enforced across `src`, `test`, and `scripts` with no exceptions.

**Refactors** — 20 functions that exceeded complexity 15 were refactored by extracting well-named helpers and flattening nested conditionals, preserving exact behavior:
- `scripts/`: `compact-test-reporter.ts`, `mutation.ts`, `mutation/isolation.ts`, `mutation/runner.ts`, `safe-upgrade.ts` (×2), `stripe-mock.ts`
- `test/`: `code-quality/detectors.ts` (×5), `checkout-pricing-consistency.test.ts`, `code-quality.test.ts`, `db/migration-restore/helpers.ts`, `fold-tree.test.ts`, `i18n-coverage.test.ts`, `square/client.test.ts`, `shared/bunny-db.test.ts`, `test-utils/fake-dom.ts`

## Verification

- `deno task lint:ci` — 1919 files, no errors/warnings.
- `deno check` over the `src`/`test`/`cli` gate — passes.
- Refactored test files run green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Chores**
  * Tightened lint enforcement for suspicious/typing checks and adjusted rule coverage for test and script files.
* **Refactor**
  * Improved mutation-run supervision and mutant evaluation, refined CLI argument parsing, and reorganized safe-upgrade execution.
  * Streamlined stripe-mock startup/retry behavior and TAP diagnostic parsing.
* **Tests**
  * Strengthened property tests with deterministic generation.
  * Enhanced code-quality scanning/tokenization utilities, database restore assertions, and other test helper ergonomics.

### CODE REVIEW

**Comment by stefan-burke** [comment]
Thanks for the review. Disposition of the 7 comments (this PR's refactors were required to preserve existing behavior, which frames several of these):

**Applied:**
- `checkout-pricing` `Rng.pick` → `readonly T[]` (widened, harmless).
- `safe-upgrade` `printUpgrades` — documented the `newPublishedAt!` invariant.

**Skipped (behavior-preserving — same as pre-PR `main`, not introduced here):**
- `mutation.ts` `applyArg` value-flag-without-token → positional: `main`'s original `parseArgs` used `arg === "--source" && next !== undefined`, so a missing value already fell through to positional. Fail-fast would be a behavior change beyond this PR's scope.
- `checkout-pricing` `randInt(0, 4)` re-evaluated in the loop condition: identical to `main` (`for (let m = 0; m < randInt(0, 4); m++)`). Sampling once would change the seeded PRNG sequence and the cases the property test exercises.

**Skipped (would regress CI):**
- `detectors.ts` `skipTypeParams` → reuse `angleDepthDelta`: the suggested form reintroduces an uncovered `return i` fall-through, which fails this repo's 100% line/branch coverage gate. Current behavior matches `main`'s original (neither special-cased `=>` in a type-param default), and no such alias exists in the scanned corpus.

**Skipped (pre-existing edge case, out of scope):**
- `detectors.ts` `skipTemplateSubstitution` not skipping comments containing `}`: this logic is unchanged from `main`; it's a latent edge case unrelated to the complexity refactor and would need its own regression test.
- `fold-tree` asserting the exact pre-fold quantity: a test-strengthening beyond the refactor's scope.

Happy to open a follow-up for the fail-fast / comment-aware-tokenizer improvements if you'd like them tracked separately.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by stefan-burke** [comment]
Applied 3 of 4 in `565ef70`:
- `compact-test-reporter` — `assignAtField` now uses a typed `Record` dispatch (matches the `VALUE_FLAGS` pattern / coding guideline).
- `stripe-mock` — only overwrite `lastStartupError` with a non-empty error, so a later empty retry can't drop the first meaningful stderr.
- `code-quality.test` — documented why `contents.get(file)!` is safe.

**Skipped:** the `rand()` `seed / 0x7fffffff` "can return 1" point. It's identical to `main` (pre-existing, not introduced by this refactor), and changing the divisor would alter the seeded PRNG sequence — i.e. change which cases the property test explores. Given the vanishing probability (one specific 31-bit seed value) and that it's out of scope for this PR, I've left it as-is rather than silently shift the deterministic corpus. Happy to fix it in a dedicated follow-up if you'd like it tracked.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: biome.json [hunk 1]
```diff
@@ -127,9 +127,7 @@
             "noExcessiveCognitiveComplexity": "off"
           },
           "suspicious": {
-            "noAssignInExpressions": "off",
-            "noConsole": "off",
-            "noImplicitAnyLet": "off"
+            "noConsole": "off"
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply biome.json:125]: Good catch — done, and enforced more broadly than a scripts-only override. `noAssignInExpressions` and `noImplicitAnyLet` are now `"error"` at the top level of `biome.json`, so they run for `src`, `test`, and `scripts` alike (not just `scripts/**`). Enabling them surfaced 6 real violations across `src`/`test`, which are fixed in this PR (assignment-in-expression rewritten to explicit assign; one implicit-any `let` given an explicit type). The scripts override now only keeps `noConsole: "off"`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/mutation.ts [hunk 2]
```diff
@@ -60,40 +60,44 @@ interface ParsedArgs {
...
+): boolean => {
+  if (arg === "--exhaustive") parsed.exhaustive = true;
+  else if (arg === "--harness") parsed.useHarness = true;
+  else if (arg === "-h" || arg === "--help") parsed.help = true;
+  else if (VALUE_FLAGS[arg] !== undefined && next !== undefined) {
+    VALUE_FLAGS[arg]?.(parsed, next);
+    return true;
+  } else positional.push(arg);
+  return false;
+};
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/mutation.ts:95]: Skipping — this preserves existing behavior. On `main`, `parseArgs` used `arg === "--source" && next !== undefined`, so a value flag with no following token already fell through to `positional`. The refactor kept that exactly. Fail-fast would be a behavior change beyond this PR's scope (which was purely reducing cognitive complexity while preserving behavior). Happy to do it as a follow-up if you'd like the UX improved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/lib/checkout-pricing-consistency.test.ts [hunk 3]
```diff
@@ -106,6 +106,79 @@ const expectConsistent = async (
...
+
+/** Insert a random modifier set for one property iteration, returning the
+ * optional-trigger add-on quantities and any promo code that was assigned. */
+const generateModifiers = async (
+  rng: Rng,
+  iter: number,
+): Promise<{ addOns: Map<number, number>; code: string | undefined }> => {
+  const addOns = new Map<number, number>();
+  let code: string | undefined;
+  for (let m = 0; m < rng.randInt(0, 4); m++) {
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/lib/checkout-pricing-consistency.test.ts:158]: Skipping — identical to `main` (`for (let m = 0; m < randInt(0, 4); m++)`), so it's pre-existing, not introduced here. Because this is a seeded property test, sampling once would change the PRNG sequence and therefore which cases the test explores. I've kept it as-is to preserve the deterministic corpus; it can be tightened in a dedicated follow-up.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/lib/code-quality/detectors.ts [hunk 4]
```diff
@@ -405,13 +405,40 @@ const NON_CALL_KEYWORDS = new Set([
...
+    const d = content[j];
+    if (d === "{") depth++;
+    else if (d === "}") depth--;
+    else if (d === '"' || d === "'" || d === "`") {
+      j = skipString(content, j);
+      continue;
+    }
+    j++;
+  }
+  return j;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/lib/code-quality/detectors.ts:439]: Skipping in this PR — the substitution-scanning logic is unchanged from `main` (the refactor only extracted it into a named helper). The `}`-in-comment case is a real but pre-existing latent edge case, out of scope for a complexity-only refactor, and the suggested fix also needs its own regression test. Worth a separate PR rather than folding it into this one.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/lib/code-quality/detectors.ts [hunk 5]
```diff
@@ -720,24 +744,56 @@ const matchBrace = (code: string, open: number): number => {
...
+  if (code[i] !== "<") return i;
+  let depth = 0;
+  for (; i < code.length; i++) {
+    if (code[i] === "<") depth++;
+    else if (code[i] === ">") {
+      depth--;
+      if (depth === 0) return i + 1;
+    }
+  }
+  return i;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/lib/code-quality/detectors.ts]: Skipping — the suggested form reintroduces a CI failure. Reusing `angleDepthDelta` here leaves an unreachable-in-practice `return i` fall-through, which this repo's 100% line/branch coverage gate rejects (that's exactly what I had to restructure away in `565ef70`'s predecessor). The `=>` handling also matches `main`'s original behavior (neither special-cased it), and no such `type A<T = () => void>` alias exists in the scanned corpus, so current behavior is correct for the inputs. A separate change could add both the `=>` handling and a covering test together.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/lib/fold-tree.test.ts [hunk 6]
```diff
@@ -62,6 +62,19 @@ const freshState = () => ({
...
+const foldOutcomeValid = (
+  error: ReturnType<typeof foldChild>,
+  recordedQty: number | undefined,
+  running: number,
+  max: number,
+): boolean =>
+  running <= max
+    ? error === null && recordedQty === running
+    : error !== null && recordedQty !== running;
+
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/lib/fold-tree.test.ts:77]: Skipping in this PR — this is a test-strengthening beyond the complexity refactor's scope (the assertion logic is equivalent to what was inlined before). It's a reasonable improvement; I'd rather land it as its own change so the invariant tightening is reviewed on its own merits.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/lib/checkout-pricing-consistency.test.ts [hunk 7]
```diff
@@ -106,6 +106,79 @@ const expectConsistent = async (
...
+
+const makeRng = (initialSeed: number): Rng => {
+  let seed = initialSeed;
+  const rand = () => {
+    seed = (seed * 1103515245 + 12345) & 0x7fffffff;
+    return seed / 0x7fffffff;
+  };
+  const pick = <T>(xs: readonly T[]): T => xs[Math.floor(rand() * xs.length)]!;
+  const randInt = (lo: number, hi: number) =>
+    lo + Math.floor(rand() * (hi - lo + 1));
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/lib/checkout-pricing-consistency.test.ts:125]: Skipping — `seed / 0x7fffffff` is identical to `main` (pre-existing, not introduced by this refactor). Changing the divisor would shift the seeded PRNG sequence, i.e. change which cases the property test explores. Given the vanishing probability (one specific 31-bit seed) and that this PR is a behavior-preserving refactor, I've left it; happy to fix it in a focused follow-up.

---
_Generated by [Claude Code](https://claude.ai/code)_
